### PR TITLE
Creating a transport action for the CoordinationDiagnosticsService

### PR DESCRIPTION
**Pull request by masseyke**
This exposes the CoordinationDiagnosticsService (#87672) through a transport action so that it can be called remotely as
part of the health API in the event that: (1) there has been no master recently, (2) there are master-eligible nodes in the cluster, (3) none are elected, and (4) the current node is not master eligible. In the diagram below, this action will be used in the 1.2.2.3 branch (it is the long arrow back to branch 1). The logic that actually uses it will be within the CoordinationDiagnosticsService, and will be in a follow-up PR. Requests for this action will be sent directly to the remote node, and the assumption is that the cluster that likely does not have a master node.
2022-04-18 Coordination health

### CODE REVIEW

**Comment by masseyke** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/action/admin/cluster/coordination/CoordinationDiagnosticsAction.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(result);
+        }
+    }
+
+    /**
+     * This transport action calls the CoordinationDiagnosticsService on a remote node.
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/coordination/CoordinationDiagnosticsAction.java:120]: nit: This doc is rather misleading to me. The transport action exposes the functionality via the transport service (as opposed to calling a service on a remote node)

**Review by andreidan — CHANGES_REQUESTED** [review]
Thanks for adding this Keith.
Left a few suggestions

File: server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java [hunk 2]
```diff
@@ -412,13 +416,47 @@ public record CoordinationDiagnosticsResult(
...
+        }
+    }
+
+    public enum CoordinationDiagnosticsStatus implements Writeable {
+        GREEN((byte) 0),
+        UNKNOWN((byte) 1),
+        YELLOW((byte) 2),
+        RED((byte) 3);
+
+        private final byte value;
```
[the remark this answers is not in the document]
Comment (masseyke) [reply server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java:439]: Huh. Somehow I didn't realize that those methods existed. Switching to use them now.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/coordination/CoordinationDiagnosticsAction.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+            this.clusterService = clusterService;
+            this.coordinator = coordinator;
+            this.masterHistoryService = masterHistoryService;
+        }
+
+        @Override
+        protected void doExecute(Task task, CoordinationDiagnosticsAction.Request request, ActionListener<Response> listener) {
+            listener.onResponse(
+                new Response(
+                    new CoordinationDiagnosticsService(clusterService, coordinator, masterHistoryService).diagnoseMasterStability(
```
[the remark this answers is not in the document]
Comment (masseyke) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/coordination/CoordinationDiagnosticsAction.java:145]: Yeah I think you're right -- I don't think there would be any thread-safety problems.

[the remark this answers is not in the document]
Comment (masseyke) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/coordination/CoordinationDiagnosticsAction.java]: And on second look, I'm not sure why I wasn't injecting that in the first place.

**Comment by masseyke** [comment]
@elasticmachine update branch

**Review by andreidan — APPROVED** [review]
LGTM, thanks for iterating on this Keith

Left a minor suggestion

File: server/src/main/java/org/elasticsearch/action/admin/cluster/coordination/CoordinationDiagnosticsAction.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+            ActionFilters actionFilters,
+            CoordinationDiagnosticsService coordinationDiagnosticsService
+        ) {
+            super(CoordinationDiagnosticsAction.NAME, transportService, actionFilters, CoordinationDiagnosticsAction.Request::new);
+            this.coordinationDiagnosticsService = coordinationDiagnosticsService;
+        }
+
+        @Override
+        protected void doExecute(Task task, CoordinationDiagnosticsAction.Request request, ActionListener<Response> listener) {
+            listener.onResponse(new Response(coordinationDiagnosticsService.diagnoseMasterStability(request.explain)));
```
[the remark this answers is not in the document]
Comment (masseyke) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/coordination/CoordinationDiagnosticsAction.java:134]: I think that's already handled for us in the parent class here right? https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/action/support/TransportAction.java#L79. I'll confirm that that is where it is called from.

[the remark this answers is not in the document]
Comment (andreidan) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/coordination/CoordinationDiagnosticsAction.java:134]: Ah you're right. TIL. Thanks for looking into this!

**Pull request merged**